File: name-score/src/main/java/com/shrader/namescore/parse/NameLoader.java
package com.shrader.namescore.parse;

import java.io.IOException;
import java.nio.CharBuffer;


public interface NameLoader<A> {
    /**
     * @param dataSource - generic source for the implementation to handle reading data from and into a CharBuffer
     * @return CharBuffer - the implementation must return a charbuffer in order to effectively decouple logic from data intake
     * @throws IOException - since we are reading from a datasource there is a chance the user could have given us incorrect data/parameters
     */
    CharBuffer load(A dataSource) throws IOException;
}

File: name-score/name-score-cli/src/main/kotlin/com/shrader/namescore/cli/parser/NameLoader.kt
package com.shrader.namescore.cli.parser

import java.io.File
import java.io.IOException


interface NameLoader<T> {
    /**
     * @param dataSource - generic source for the implementation to handle reading data from and into a CharBuffer
     * @return List<String> - the implementation must return a List<String> in order to effectively decouple logic from data intake
     * @throws IOException - since we are reading from a datasource there is a chance the user could have given us incorrect data/parameters
     */
    @Throws(IOException::class)
    fun load(dataSource: File, delimiter: String): MutableList<String>
}
